### ISSUE TITLE
Kick-Start

### ISSUE THREAD

**Issue by ezorita** [palicode/nucleotid#1]
## Description
Preliminary definitions to kick-start the project.

## Subtasks
- [x] Etor joins github repo and waffle.
- [x] Create organization where two repos (backend and frontend) will be maintained.
- [ ] Agree on organization name.
- [ ] Agree on git workflow (Issue templates, waffle bot, PR, etc.).
- [ ] Etor reads current implementation dependencies and decides which are missing and which should be removed.
- [ ] Agree on a code format and set lint up.
- [x] Choose database server: **PostgreSQL**
- [ ] Front-end skeleton (reactjs, some CSS magic (?), tests...).
- [ ] Set up test environment.
- [ ] Set up CI (Travis).
- [ ] Configure dev server (heroku/Docker).

## Acceptance criteria
- [ ] Minimal front-end demo which uses some dummy DB data.
- [ ] Code is passed through CI and tests before merging to master.
- [ ] Working dev server.

**Comment by espinacs** [comment]
Hi @ezorita,

Regarding this issue, and quite before starting with this tasks, i think we should decide whether to split front/back apps into two separate repos, or keep it all together, as it is now.

First option would complicate thinks at the beginning, and maybe make us loose track of each project while working on the other, but at the same time every side could have better maintenance and performance, and dev environments would become easier to install and launch.

On the other hand, keeping it all together would be an option, but in that case i'd go for a serverless (https://medium.com/@gethylgeorge/setting-up-a-simple-graphql-server-with-node-express-and-mongoose-ff8a1071af53) app, as we discussed yesterday. Also, I think a non-relational database is an overkiller for our goal.

Let me know your thoughts about this

**Comment by ezorita** [comment]
Regarding the DB, what server do you suggest to use? Having JSON outputs simplifies the code quite a bit.

**Comment by espinacs** [comment]
- I think nowadays the best **relational db** is PostgreSQL as performance regards, especially with complex queries, and syntax is not too different from mySQL. Also, although both store JSON, PostgreSQL can index it. Anyhow, in my opinion MySQL should be enough for our purpose.

- Regarding to the **JSON usage**, keeping the node API pure REST shouldn't be a problem, doing like this
```
function getAllPuppies(req, res, next) {
  db.any('select * from pups')
    .then(function (data) {
      res.status(200)
        .json({
          status: 'success',
          data: data,
          message: 'Retrieved ALL puppies'
        });
    })
    .catch(function (err) {
      return next(err);
    });
}
```

- Ok, let's keep it simple with a **monorepo** and **without graphQL**

**Comment by espinacs** [comment]
Also, and in order to split sides but keeping it on the same repo,  i'd go for 2 folders, `/client` and `/server` (what nowadays is `/backend`), and each one should have its own package.json and everything.

In this case, i think we should create an specific task (with subtasks) in order to implement a first approach of the frontend universe: the react app with it's unitary tests and e2e tests, specific javascript es6 linter, it's own css framework (thinking about Atlaskit), etc.

Looking forward to meet again.

**Comment by espinacs** [comment]
By the way, I'd seriously recommend to split this project into 2 separate repos. Some references:

- https://quickleft.com/blog/six-reasons-we-split-front-end-and-back-end-code-into-two-git-repositories-working-with-github-repositories/
- https://medium.com/@joeyvmason/the-front-end-deserves-its-own-repository-86fe382b7d37
- https://medium.com/@MentallyFriendly/split-stack-development-a-model-for-modern-applications-d7b9abb47bd5

If your concerns are about PM integration, i've read waffle can stand it: https://help.waffle.io/waffle-boards/creating-project-boards/how-do-i-add-multiple-repositories-to-a-single-board

**Comment by ezorita** [comment]
- At this point is almost impossible to predict where the DB architecture will cause bottlenecks. If your gut feeling is that **PostgreSQL** is the most efficient, let's use it for now.
- Good, after reading the blog posts and solving the multi-board issue, it makes a lot of sense to have two repos.
- Maybe we should create the repos in an organization account, instead of using mine. Any suggestions for the organization name? The first time we talked about this was in the "FM barra"... _barrabits, barracode, ginbits, pal i code, pal i bits_... Show your pun skills.

**Comment by ezorita** [comment]
Well, until we find a better name, I transferred the ownership of the repo to an organization named **palicode**. Perhaps we should rename the repo as well to something like **nucleotid-back** and create a new repo **nucleotid-front**?

**Comment by espinacs** [comment]
Hey @ezorita, good work.

However, i thought we'd use *nucleotid* to sign our products, instead of *palicode* or anything else. Then, projects could be all named using any universe of words, like sega characters, vallès towns, human towers positions, pg players, etc.

Anyhow, i've been working already on a separate frontend project, so as long as this name thing is decided i'll upload it.

**Comment by ezorita** [comment]
Yes, but _nucleotid_ is not available, so I thought we could define our organization name (analogous to a company) name and then have _nucleotid_ as a product.

In this case, _palicode_ is (for now) the company name and one of our (many) products is _nucleotid_, hence it would make more sense to call them _nucleotid-front_ and _nucleotid-back_.

Just give them dummy names for now, we can always rename later.

**Comment by espinacs** [comment]
Great. I've already created a new repo and pushed some initial configuration

I've created a new task (https://waffle.io/palicode/nucleotid/cards/5b9e8843194e3600395119c4) in order to set to this repo all the config things we can need (test frameworks, deployment options, linting, structure...).

I'd like to finish this in a couple days, and then we could meet and discuss it, so you are ok with all.

**Comment by ezorita** [comment]
Very good! I will try to Dockerize the backend, so that we can deploy a dev environment anywhere in no time. We can meet again on Thursday to review the progress and do some frikin amazin live hacking.

**Comment by espinacs** [comment]
Deal.

**Comment by ezorita** [comment]
I have renamed the backend repository to **nucleotid-back**, so please update your remote config:

- [ ] `git remote set-url origin git@github.com:palicode/nucleotid-back`

**Comment by ezorita** [comment]
After some hacking I managed to configure a preliminary **Dockerized environment for development**, which I documented in [the wiki](https://github.com/palicode/nucleotid-back/wiki/Dev-environment-with-Docker-Compose). 

Can you try to follow the steps and tell me if it works for you?

**Comment by espinacs** [comment]
I'm trying to follow the wiki steps, but having (too much) problems, some involving code. Should I open a new issue instead?

**Comment by ezorita** [comment]
Yes, let's discuss this in a new Issue.

**Comment by ezorita** [comment]
btw the server is not yet working properly, the goal is that you get the database (**db** service) and the server (**node** server) running properly (even though you'll face some JavaScript error) after calling `npm run docker`, i.e. that you see the output from both.

**Comment by espinacs** [comment]
Trying this tomorrow again with a fresh setup, and reporting issues. Good night!

**Comment by ezorita** [comment]
FYI these are the lines you should see (plus other error crap):
```
(...)
db_1    | 2018-09-17 17:36:19.621 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1    | 2018-09-17 17:36:19.621 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1    | 2018-09-17 17:36:19.713 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1    | 2018-09-17 17:36:19.965 UTC [24] LOG:  database system was shut down at 2018-09-17 17:35:36 UTC
db_1    | 2018-09-17 17:36:20.043 UTC [1] LOG:  database system is ready to accept connections
node_1  | npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
node_1  | npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
node_1  | 
node_1  | audited 1531 packages in 3.617s
node_1  | found 0 vulnerabilities
node_1  | 
node_1  | 
node_1  | > nucleotid@0.1.0 devstart /app
node_1  | > nodemon ./src/server
node_1  | 
node_1  | [nodemon] 1.18.3
node_1  | [nodemon] to restart at any time, enter `rs`
node_1  | [nodemon] watching: *.*
node_1  | [nodemon] starting `node ./src/server`
node_1  | module.js:550
node_1  |     throw err;
node_1  |     ^
node_1  | 
node_1  | Error: Cannot find module 'mongoose'
node_1  |     at Function.Module._resolveFilename (module.js:548:15)
node_1  |     at Function.Module._load (module.js:475:25)
node_1  |     at Module.require (module.js:597:17)
node_1  |     at require (internal/module.js:11:18)
node_1  |     at Object.<anonymous> (/app/src/models/user.js:1:78)
node_1  |     at Module._compile (module.js:653:30)
node_1  |     at Object.Module._extensions..js (module.js:664:10)
node_1  |     at Module.load (module.js:566:32)
node_1  |     at tryModuleLoad (module.js:506:12)
node_1  |     at Function.Module._load (module.js:498:3)
node_1  | [nodemon] app crashed - waiting for file changes before starting...
```

**Comment by ezorita** [comment]
I added some more info to the [Wiki](https://github.com/palicode/nucleotid-back/wiki), in case you want to set up the dev environment @ home. The current code is working, you only need to set your own Google OAUTH2 keys to make it work.
